### ISSUE TITLE
A setting to disable default mappings

### ISSUE THREAD

**Issue by Melkster** [vim-pandoc/vim-markdownfootnotes#20]
By default, this plugin hijacks `<leader>f` and `<leader>r` in normal mode _and_ insert mode. The insert mode mapping makes this plugin unusable for anyone who has for instance has `<leader>` mapped to `<space>`.

There should definitely be a setting for disabling default mappings.

**Comment by alerque** [comment]
There should already be a way to handle this. If you create your own bindings to the plugin's functions first, the plugin won't add it's default ones. It only adds them if it finds that there are no existing bindings. There are two commands in two modes for a total of four bindings that get created in none others exist. The defaults are:

```viml
imap <buffer> <Leader>f <Plug>AddVimFootnote
nmap <buffer> <Leader>f <Plug>AddVimFootnote
imap <buffer> <Leader>r <Plug>ReturnFromFootnote
nmap <buffer> <Leader>r <Plug>ReturnFromFootnote
```

If you map those functions to your own keys (you can use anything it doesn't have to involve the leader at all) you should get the effect you want. If you don't want a binding at all bind it to something impossible.

I agree that is less than ideal and a documented way to suppress this (particularly the insert mode bindings) might be useful. I'd be happy to review and merge any contribution that took this on.

**Comment by alerque** [comment]
Also if you could review #17 that might be useful. I proposed some changes particularly to the insert maps and simplifying to have fewer bindings by default, but we never really agreed on whether my proposal was an improvement. Feedback would be great.

**Comment by m040601** [comment]
> ... **If** you create **your own bindings** to the plugin's functions first, the plugin **won't add its default ones**. It only adds them if it finds that there are no existing bindings. 

Yes, this does work in fact. 
Could you add this information the the markdownfootnotes.txt in the docs ?

For  example if I choose to "reserve" \n to this plugin and add this to my vim/nvim config:

```
imap <buffer> <Leader>ni <Plug>AddVimFootnote
nmap <buffer> <Leader>ni <Plug>AddVimFootnote
imap <buffer> <Leader>nr <Plug>ReturnFromFootnote
nmap <buffer> <Leader>nr <Plug>ReturnFromFootnote

```

Just a small question about the "noremap" thing. Pardon my vim ignorance.

If I instead do:

```
inoremap <buffer> <Leader>ni <Plug>AddVimFootnote
nnoremap <buffer> <Leader>ni <Plug>AddVimFootnote
inoremap <buffer> <Leader>nr <Plug>ReturnFromFootnote
nnoremap <buffer> <Leader>nr <Plug>ReturnFromFootnote

```

this doesntt work anymore. Why ? I've always been told that it's "best" to choose the "noremap" thing.

**Comment by Melkster** [comment]
> There should already be a way to handle this. If you create your own bindings to the plugin's functions first, the plugin won't add its default ones

@alerque How would I go about to simply disable the insert mappings? Could this be done without remapping `<Plug>AddVimFootnote` and `<Plug>ReturnFromFootnote` to some obscure key in insert mode?

**Comment by alerque** [comment]
No, I don't think that can be done right now. It should be possible (hence this issue being open!).

**Comment by Konfekt** [comment]
I currently work around it by

```vim
imap <SID>(DisableAddVimFootnote)     <Plug>AddVimFootnote
imap <SID>(DisableReturnFromFootnote) <Plug>ReturnFromFootnote
nmap <SID>(DisableReturnFromFootnote) <Plug>ReturnFromFootnote

augroup epnMarkdownFootnote
  autocmd!
  autocmd FileType markdown inoreabbrev afn <c-o>:<c-u>exe "normal \<Plug>AddVimFootnote"<cr>
augroup end

```

**Comment by Melkster** [comment]
Here's another workaround. Remap it to `<F13>` and `<F14>` in insert mode (valid keys that aren't on most keyboards):

```vim
imap <buffer> <F13> <Plug>AddVimFootnote
imap <buffer> <F14> <Plug>ReturnFromFootnote
```

I would prefer a builtin feature to simply disable the insert mode mapping, but this seems to do the trick.

**Comment by falko-strenzke** [comment]
Please note that the imap mappings leads to the following unintended and highly undesirable behaviour:
Typing "\r" in a Markdown buffer causes that buffer to be closed immediately with Vim exiting if that was the only open buffer. This will lead to loss of unsaved data.
